msprime/simulations.py

```diff
@@ -963,6 +964,69 @@ def mean_recombination_rate(self):
 
         return np.average(rates, weights=weights)
 
+    def slice(self, start=None, end=None, remove_flanks=False):
+        """
+        Returns a subset of this recombination map between the specified end
+        points. If start is None, it defaults to 0. If end is None, it defaults
+        to the end of the map. If remove_flanks is True, remove the flanking
+        zero recombination rate regions such that the sequence length of the
+        new recombination map is end - start.
+        """
+        positions = self.get_positions()
+        rates = self.get_rates()
+
+        if start is None:
+            i = 0
+            start = 0
+        if end is None:
+            end = positions[-1]
+            j = len(positions)
+
+        # TODO: Is there interest in supporting negative indexes? Eg. make
+        #       recomb_map[:-x] mean end=positions[-1] - x
+        if (start < 0 or end < 0 or start > positions[-1] or end > positions[-1]
+           or start > end):
+            raise IndexError(f"Invalid slice: start={start}, end={end}")
+
+        if start != 0:
+            i = bisect.bisect_left(positions, start)
+            if start < positions[i]:
+                i -= 1
+        if end != positions[-1]:
+            j = bisect.bisect_right(positions, end, lo=i)
+
+        new_positions = positions[i:j]
+        new_rates = rates[i:j]
+        new_positions[0] = start
+        if end > new_positions[-1]:
+            new_positions.append(end)
+            new_rates.append(0)
+        else:
+            new_rates[-1] = 0
+        if remove_flanks:
+            new_positions = [pos-start for pos in new_positions]
+        else:
+            if new_positions[0] != 0:
+                if new_rates[0] == 0:
+                    new_positions[0] = 0
+                else:
+                    new_positions.insert(0, 0)
+                    new_rates.insert(0, 0.0)
+            if new_positions[-1] != positions[-1]:
+                new_positions.append(positions[-1])
+                new_rates.append(0)
+        return self.__class__(
+                new_positions, new_rates, discrete=self.discrete)
+
+    def __getitem__(self, key):
```

Comment:
Very nice idea, I like it! What happens when we give non-integer arguments like ``m[0.5:0.9]``? Very cool if it works!

Reply:
I didn't try it before, but it seems to work just fine.
```
>>> m = msprime.RecombinationMap.read_hapmap("chr1_average_canFam3.1.txt")
>>> m[10e6:11e6].get_positions()
[0.0, 8027.0, 683500.0, 780593.0, 807939.0, 934077.0, 961544.0, 1000000.0]
>>> m[10e6+0.001:11e6+0.7].get_positions()
[0.0, 8026.998999999836, 683499.9989999998, 780592.9989999998, 807938.9989999998, 934076.9989999998, 961543.9989999998, 1000000.6989999991]
```

Reply:
Beautiful! OK, let's write a few test cases for this behaviour and merge. This is awesome.